File: src/main/java/org/revcloud/vader/matchers/DateMatchers.java
package org.revcloud.vader.matchers;

import io.vavr.Function2;
import lombok.experimental.UtilityClass;

import java.util.Date;

@UtilityClass
public class DateMatchers {
    @SuppressWarnings({"java:S1874"})
    public static Function2<Object, Object, Boolean> isEqualToDayOfDate() {
        return (day, date) -> {
            if (day == null && date == null) {
                return false;
            }
            if (!(day instanceof Integer) || !(date instanceof Date)) {
                return false;
            }
            return (Integer) day == ((Date) date).getDate();
        };
    }

    public static Function2<Object, Object, Boolean> isOnOrBeforeIfBothArePresent() {
        return (date1, date2) -> {
            if (date1 == null || date2 == null) {
                return true;
            }
            if (!(date1 instanceof Date) || !(date2 instanceof Date)) {
                return false;
            }
            return date1.equals(date2) || ((Date) date1).before((Date) date2);
        };
    }

    public static Function2<Object, Object, Boolean> isBeforeIfBothArePresent() {
        return (date1, date2) -> {
            if (date1 == null && date2 == null) {
                return true;
            }
            if (!(date1 instanceof Date) || !(date2 instanceof Date)) {
                return false;
            }
            return ((Date) date1).before((Date) date2);
        };
    }
}

File: src/main/java/org/revcloud/vader/matchers/DateMatchers.kt
@file:JvmName("DateMatchers")

package org.revcloud.vader.matchers

import io.vavr.Function2
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeMatcher
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Date

val isOnOrBeforeIfBothArePresent = Function2 { date1: Any?, date2: Any? ->
  when {
    date1 == null || date2 == null -> true
    date1 !is Date || date2 !is Date -> false
    else -> date1 == date2 || date1.before(date2)
  }
}

val isBeforeIfBothArePresent = Function2 { date1: Any?, date2: Any? ->
  when {
    date1 == null && date2 == null -> true
    date1 !is Date || date2 !is Date -> false
    else -> date1.before(date2)
  }
}

@Suppress("DEPRECATION")
val isEqualToDayOfDate = Function2 { day: Any?, date: Any? ->
  when {
    day == null && date == null -> false
    day !is Int || date !is Date -> false
    else -> day == date.date
  }
}

@get:JvmName("ISO8601DateFormat")
val ISO8601DateFormat: Matcher<Any?> = object : TypeSafeMatcher<Any?>() {

  override fun describeTo(description: Description?) {
    description?.appendText("ISO 8601 format")
  }

  override fun matchesSafely(date: Any?): Boolean =
    date is String && isValidIS0LocalDateFormat(date)
}

private fun isValidIS0LocalDateFormat(dateAsString: String): Boolean =
  try {
    parseISOLocalDate(dateAsString)
    true
  } catch (e: Exception) {
    false
  }

private fun parseISOLocalDate(date: String): Date? =
  Date.from(
    LocalDate
      .parse(date.trim(), DateTimeFormatter.ISO_LOCAL_DATE)
      .atStartOfDay(ZoneOffset.UTC).toInstant()
  )
